pip/pip-439.md

```diff
@@ -0,0 +1,573 @@
+        return delegate.getKey();
+    }
+
+    @Override
+    public T getValue() {
+        return delegate.getValue();
+    }
+
+    @Override
+    public Optional<Long> getEventTime() {
+        return delegate.getEventTime();
+    }
+
+    @Override
+    public Optional<String> getPartitionId() {
+        return delegate.getPartitionId();
+    }
+
+    @Override
+    public Optional<String> getTopicName() {
+        return delegate.getTopicName();
+    }
+
+    @Override
+    public Schema<T> getSchema() {
+        return delegate.getSchema();
+    }
+
+    @Override
+    public Optional<Long> getRecordSequence() {
+        return delegate.getRecordSequence();
+    }
+
+    @Override
+    public Map<String, String> getProperties() {
+        return delegate.getProperties();
+    }
+}
+```
+
+### Modifications to JavaInstanceRunnable
+
+The JavaInstanceRunnable class needs to be updated to handle auto transactions:
+
+```java
+public class JavaInstanceRunnable implements AutoCloseable, Runnable {
+    // Existing fields...
+
+    @Override
+    public void run() {
+        try {
+            setup();
+
+            while (true) {
+                currentRecord = readInput();
+
+                // increment number of records received from source
+                stats.incrTotalReceived();
+
+                // If auto transactions are enabled, wrap the function execution in a transaction
+                if (instanceConfig.getFunctionDetails().getAutoTransactionsEnabled()) {
+                    processWithTransaction();
+                } else {
+                    processNormally();
+                }
+
+                if (deathException != null) {
+                    throw deathException;
+                }
+            }
+        } catch (Throwable t) {
+            // Existing error handling...
+        } finally {
+            close();
+        }
+    }
+
+    private void processNormally() {
+        // Existing logic for non-transactional execution
+        Thread currentThread = Thread.currentThread();
+        Consumer<Throwable> asyncErrorHandler = throwable -> currentThread.interrupt();
+
+        // set last invocation time
+        stats.setLastInvocation(System.currentTimeMillis());
+
+        // start time for process latency stat
+        stats.processTimeStart();
+
+        // process the message
+        Thread.currentThread().setContextClassLoader(functionClassLoader);
+        JavaExecutionResult result = javaInstance.handleMessage(
+                currentRecord,
+                currentRecord.getValue(),
+                this::handleResult,
+                asyncErrorHandler);
+        Thread.currentThread().setContextClassLoader(instanceClassLoader);
+
+        // register end time
+        stats.processTimeEnd();
+
+        if (result != null) {
+            // process the synchronous results
+            handleResult(currentRecord, result);
+        }
+    }
+
+    private void processWithTransaction() {
```

Comment:
This doesn't look right. In Pulsar, you can asynchronously handle messages so that you have more than one message in flight at a time. It's important for performance at scale since it enables "pipelining".
That's why it's better to handle use the existing solution and refactor it to support transactions.